File: examples/src/main/scala/portals/examples/PortalPingPong.scala
package portals.examples

import scala.annotation.experimental
import scala.concurrent.Await

import portals.*

import portals.api.dsl.DSL
import portals.api.dsl.ExperimentalDSL
@experimental
@main def PortalPingPong(): Unit =
  import portals.api.dsl.DSL.*

  import portals.api.dsl.ExperimentalDSL.*

  val app = PortalsApp("PortalPingPong") {
    sealed trait PingPong
    case class Ping(x: Int) extends PingPong
    case class Pong(x: Int) extends PingPong

    val generator = Generators.fromList(List(1024 * 1024))

    val portal = Portal[Ping, Pong]("portal")

    val replier = Workflows[Nothing, Nothing]("replier")
      .source(Generators.empty.stream)
      .replier[Nothing](portal) { _ => () } { case Ping(x) =>
        reply(Pong(x - 1))
      }
      .sink()
      .freeze()

    val asker = Workflows[Int, Int]("asker")
      .source(generator.stream)
      .recursiveAsker[Int](portal) { self => x =>
        val future: Future[Pong] = ask(portal)(Ping(x))
        future.await {
          ctx.emit(future.value.get.x)
          if future.value.get.x > 0 then self(future.value.get.x)
        }
      }
      .filter(_ % 1024 == 0)
      .logger()
      .sink()
      .freeze()
  }

  val system = Systems.test()

  system.launch(app)

  system.stepUntilComplete()

  system.shutdown()

File: portals-js/src/main/resources/PortalPingPong.js
const app = new PortalsJS.PortalsApp("PortalPingPong", () => {
    class PingPong {}
    class Ping extends PingPong {
        constructor(x) {
            super();
            this.x = x;
        }
    }
    class Pong extends PingPong {
        constructor(x) {
            super();
            this.x = x;
        }
    }

    const generator = Generators.fromList(Array(1024 * 128).fill(0));

    const portal = new Portal("portal", Ping, Pong);

    const replier = new Workflows("replier", null, null)
        .source(Generators.empty.stream)
        .replier(portal, () => {})((_, message) => {
            if (message instanceof Ping) {
                return new Pong(message.x - 1);
            }
        })
        .sink()
        .freeze();

    const asker = new Workflows("asker", Number, Number)
        .source(generator.stream)
        .recursiveAsker(portal)((self, x) => {
            const future = ask(portal)(new Ping(x));
            future.await(() => {
                const value = future.value.get;
                ctx.emit(value.x);
                if (value.x > 0) {
                    self(value.x);
                }
            });
        })
        .filter(value => value % 1024 === 0)
        .logger()
        .sink()
        .freeze();
});

const system = Systems.interpreter();

system.launch(app);

system.stepUntilComplete();

system.shutdown();
